### PR TITLE
Add `round` and `sumAll` functions to `BigDecimal`

### PR DESCRIPTION
**Pull request by jrudder**
Add `round` and `sumAll` functions to `BigDecimal`

## Type

- [ ] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

The `round` function supports rounding at an arbitrary scale/position and the following 10 rounding modes:
- `ceil`: round towards positive infinity
- `floor`: round towards negative infinity
- `to-zero`: round towards zero
- `from-zero`: round away from zero
- `half-ceil`: round to the nearest neighbor; if equidistant round towards positive infinity
- `half-floor`: round to the nearest neighbor; if equidistant round towards negative infinity
- `half-to-zero`: round to the nearest neighbor; if equidistant round towards zero
- `half-from-zero`: round to the nearest neighbor; if equidistant round away from zero
- `half-even`: round to the nearest neighbor; if equidistant round to the neighbor with an even digit at the specified scale
- `half-odd`: round to the nearest neighbor; if equidistant round to the neighbor with an odd digit at the specified scale

The default scale is `0` (round to integer), and the default mode is `half-from-zero`.

The `sumAll` function mirrors that of `BigInt.sumAll`.

## Related

_None_

### CODE REVIEW

**Comment by fubhy** [comment]
Thanks! Support for rounding has been sorely missing.

**Review by fubhy — CHANGES_REQUESTED** [review]
Can you make these rounding functions `dual` (data last)

**Comment by jrudder** [comment]
@fubhy Thank you for looking into this so quickly.

I have now:

- updated the functions to use `dual`
- cleaned up the JSDoc examples
- changed `digitAt` to use `self` (like the existing `normalize` and other functions, including the others I added)

File: packages/effect/src/BigDecimal.ts [hunk 1]
```diff
@@ -1126,3 +1126,234 @@ export const isNegative = (n: BigDecimal): boolean => n.value < bigint0
...
+/**
+ * Calculate the ceiling of a `BigDecimal` at the given scale.
+ *
+ * @example
+ * ```ts
+ * import * as assert from "node:assert"
+ * import { ceil, unsafeFromString } from "effect/BigDecimal"
+ *
+ * assert.deepStrictEqual(ceil(unsafeFromString("145"), -1), unsafeFromString("150"))
+ * assert.deepStrictEqual(ceil(-1)(unsafeFromString("145")), unsafeFromString("150"))
```
[the remark this answers is not in the document]
Comment (fubhy) [reply packages/effect/src/BigDecimal.ts]: Took the liberty to take these out.

File: packages/effect/src/BigDecimal.ts [hunk 2]
```diff
@@ -1126,3 +1126,234 @@ export const isNegative = (n: BigDecimal): boolean => n.value < bigint0
...
+ * import * as assert from "node:assert"
+ * import { unsafeFromString, sumAll } from "effect/BigDecimal"
+ *
+ * assert.deepStrictEqual(
+ *  sumAll(
+ *    [
+ *      unsafeFromString("2"),
+ *      unsafeFromString("3"),
+ *      unsafeFromString("4")
+ *    ]), unsafeFromString("9"))
```
[the remark this answers is not in the document]
Comment (fubhy) [reply packages/effect/src/BigDecimal.ts]: Took the liberty to adjust this.

File: packages/effect/src/BigDecimal.ts [hunk 3]
```diff
@@ -1126,3 +1126,234 @@ export const isNegative = (n: BigDecimal): boolean => n.value < bigint0
...
+ * `ceil`: round towards positive infinity
+ * `floor`: round towards negative infinity
+ * `to-zero`: round towards zero
+ * `from-zero`: round away from zero
+ * `half-ceil`: round to the nearest neighbor; if equidistant round towards positive infinity
+ * `half-floor`: round to the nearest neighbor; if equidistant round towards negative infinity
+ * `half-to-zero`: round to the nearest neighbor; if equidistant round towards zero
+ * `half-from-zero`: round to the nearest neighbor; if equidistant round away from zero
+ * `half-even`: round to the nearest neighbor; if equidistant round to the neighbor with an even digit
+ * `half-odd`: round to the nearest neighbor; if equidistant round to the neighbor with an odd digit
```
[the remark this answers is not in the document]
Comment (jrudder) [reply packages/effect/src/BigDecimal.ts:1246]: I wanted to avoid ambiguity of "up", which can mean "toward infinity" (as in Rust) or "away from zero" (as in Java and decimal.js). Similar for "down".

**References:**

[Rust round module](https://docs.rs/libmath/latest/math/round/index.html):
- `ceil`: "Round up"
- `floor`: "Round down"
- `half_away_from_zero`: "Round half away from zero"
- `half_down`: "Round half down"
- `half_to_even`: "Round half to nearest even number"
- `half_to_odd`: "Round half to nearest odd number"
- `half_towards_zero`: "Round half towards zero"
- `half_up`: "Round half up"
- `stochastic`: "Round half randomly up or down"

[Java RoundingMode](https://docs.oracle.com/en/java/javase/24/docs/api/java.base/java/math/RoundingMode.html):
- `CEILING`: "Rounding mode to round towards positive infinity"
- `DOWN`: "Rounding mode to round towards zero"
- `FLOOR`: "Rounding mode to round towards negative infinity"
- `HALF_DOWN`: "...unless both neighbors are equidistant, in which case round down."
- `HALF_EVEN`: "...in which case, round towards the even neighbor"
- `HALF_UP`: "...in which case round up"
- `UNNECESSARY`: "...no rounding is necessary"
- `UP`: "Rounding mode to round away from zero"

[decimal.js rounding modes](https://mikemcl.github.io/decimal.js/#modes):
- `ROUND_UP`: "Rounds away from zero"
- `ROUND_DOWN`: "Rounds towards zero"
- `ROUND_CEIL`: "Rounds towards Infinity"
- `ROUND_FLOOR`: "Rounds towards -Infinity"
- `ROUND_HALF_UP`: "...away from zero"
- `ROUND_HALF_DOWN`: "...towards zero"
- `ROUND_HALF_EVEN`: "...towards even neighbour"
- `ROUND_HALF_CEIL`: "...towards Infinity"
- `ROUND_HALF_FLOOR`: "...towards -Infinity"
- `EUCLID`: "Not a rounding mode"

File: packages/effect/test/BigDecimal.test.ts [hunk 4]
```diff
@@ -361,6 +361,181 @@ describe("BigDecimal", () => {
...
+          `Expected ${value} to round to ${halfToZero} with mode halfToZero, scale ${scale}`
+        )
+        assertEquals(
+          BigDecimal.round(value, { mode: "half-from-zero", scale }),
+          halfFromZero,
+          `Expected ${value} to round to ${halfFromZero} with mode halfFromZero, scale ${scale}`
+        )
+      }
+    }
+  })
```
[the remark this answers is not in the document]
Comment (jrudder) [reply packages/effect/test/BigDecimal.test.ts:530]: Absolutely. I pulled this from the [Unicode ICU rounding modes page](https://htmlpreview.github.io/?https://github.com/unicode-org/icu-demos/blob/main/roundmode/round.html) to quickly get a test-set going, but I can certainly swap it out with hand-picked values that focus on transition spots and edge-cases.

File: packages/effect/src/BigDecimal.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (fubhy) [reply packages/effect/src/BigDecimal.ts:1246]: Ok, that makes sense. Ty.

File: packages/effect/test/BigDecimal.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (jrudder) [reply packages/effect/test/BigDecimal.test.ts]: I've swapped out the table tests in commit [b4d1d24](https://github.com/Effect-TS/effect/pull/4920/commits/b4d1d2442eb2bdd366c70b997eacb6310a3789fa)
